File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0874_Walking_Robot_Simulation.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashSet;
import java.util.Set;

public class Question_0874_Walking_Robot_Simulation {
    public int robotSim(int[] commands, int[][] obstacles) {
        Set<String> obs = new HashSet<>();
        for (int i = 0; i < obstacles.length; i++) {
            obs.add(help(obstacles[i][0], obstacles[i][1]));
        }
        int res = 0, dir = 0, x = 0, y = 0;
        int[][] ds = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for (int i = 0; i < commands.length; i++) {
            if (commands[i] == -2) {
                dir--;
            } else if (commands[i] == -1) {
                dir++;
            } else {
                for (int j = 0; j < commands[i]; j++) {
                    String pos = help(x + ds[dir][0], y + ds[dir][1]);
                    if (obs.contains(pos)) {
                        break;
                    }
                    x += ds[dir][0];
                    y += ds[dir][1];
                }
                res = Math.max(res, x * x + y * y);
            }
            if (dir == -1) {
                dir = 3;
            }
            if (dir == 4) {
                dir = 0;
            }
        }
        return res;
    }

    private String help(int x, int y) {
        return x + "#" + y;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0874_Walking_Robot_Simulation.cc
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
 public:
  int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
    unordered_set<string> obs;
    for (int i = 0; i < obstacles.size(); i++) {
      obs.insert(help(obstacles[i][0], obstacles[i][1]));
    }
    int res = 0, dir = 0, x = 0, y = 0;
    vector<vector<int>> ds = {{0,1}, {1,0}, {0,-1}, {-1,0}};
    for (int i = 0; i < commands.size(); i++) {
      if (commands[i] == -2) {
        dir--;
      } else if (commands[i] == -1) {
        dir++;
      } else {
        for (int j = 0; j < commands[i]; j++) {
          string pos = help(x + ds[dir][0], y + ds[dir][1]);
          if (obs.count(pos)) {
            break;
          }
          x += ds[dir][0];
          y += ds[dir][1];
        }
        res = max(res, x * x + y * y);
      }
      if(dir == -1) {
        dir = 3;
      }
      if(dir == 4) {
        dir = 0;
      }
    }
    return res;
  }

 private:
  string help(int x, int y) {
    return to_string(x) + "#" + to_string(y);
  }
};
